File: solutions/Leetcode_17/leetcode_17.py
from typing import List


class Solution1:
    def letter_combinations(self, digits: str) -> List[str]:
        if digits == '':
            return []
        letters = ['abc', 'def', 'ghi', 'jkl', 'mno', 'pqrs', 'tuv', 'wxyz']
        combinations = []

        def dfs(remain_digits, string):
            if remain_digits == '':
                combinations.append(string)
            else:
                for s in letters[int(remain_digits[0])-2]:
                    dfs(remain_digits[1:], string+s)

        dfs(digits, '')
        return combinations


class Solution2:
    def letter_combinations(self, digits):
        if digits == '':
            return []
        letters = ['abc', 'def', 'ghi', 'jkl', 'mno', 'pqrs', 'tuv', 'wxyz']
        combinations = ['']
        for i in range(len(digits)):
            temp = []
            while combinations:
                string = combinations.pop()
                for s in letters[int(digits[i]) - 2]:
                    temp.append(string + s)
            combinations = temp
        return combinations


if __name__ == '__main__':
    digits = '23'
    ans = Solution2().letter_combinations(digits)
    print(ans)

File: solutions/Leetcode_17/leetcode_17.cc
#include <string>
#include <vector>
#include <iostream>

using namespace std;

class Solution1
{
public:
    vector<string> letterCombination(string digits)
    {
        vector<string> combinations;
        if (digits.size() == 0)
            return combinations;
        string str;
        const vector<string> letters{"abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        dfs(str, digits, 0, combinations, letters);
        return combinations;
    }

    void dfs(string str, const string &digits, int id, vector<string> &combinations, const vector<string> &letters)
    {
        if (id == digits.size())
            combinations.push_back(str);
        else
        {
            for (auto s : letters[digits[id] - '0' - 2])
            {
                dfs(str + s, digits, id + 1, combinations, letters);
            }
        }
    }
};

class Solution2
{
public:
    vector<string> letterCombination(string digits)
    {
        vector<string> combinations;
        if (digits.size() == 0)
            return combinations;
        const vector<string> letters{"abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        combinations.push_back("");

        for (auto s : digits)
        {
            vector<string> temp;
            for (auto str : combinations)
            {
                for (auto x : letters[s - '0' - 2])
                    temp.push_back(str + x);
            }
            combinations.swap(temp);
        }
        return combinations;
    }
};